pycloudlib/lxd/tests/test_instance.py

```diff
@@ -69,3 +72,81 @@ def test_file_push_with_agent_on_xenial_machine(
         instance.push_file("/some/file", "/some/local/file")
         assert self._missing_agent_msg in caplog.text
         assert m_subp.call_count == 1
+        expected_msg = (
+            "Many Xenial images do not support `lxc file push` due to missing"
+            " lxd-agent: you may see unavoidable failures.\n"
+            "See https://github.com/canonical/pycloudlib/issues/132 for"
+            " details."
+        )
+        assert expected_msg in caplog.messages
+        assert m_subp.call_count == 1
+
+
+class TestIP:
+    """Tests covering pycloudlib.lxd.instance.Instance.ip."""
+
+    @pytest.mark.parametrize(
+        "stdout,stderr,return_code,sleeps,expected",
+        (
+             (
+                  "unparseable", "", 0, 150,
+                  TimeoutError(
+                      "Unable to determine IP address after 150 retries."
+                      " exit:0 stdout: unparseable stderr: "
+                  )
+             ),
+             (    # retry on non-zero exit code
+                  "10.0.0.1 (eth0)", "", 1, 150,
+                  TimeoutError(
+                      "Unable to determine IP address after 150 retries."
+                      " exit:1 stdout: 10.0.0.1 (eth0) stderr: "
+                  )
+             ),
+             (    # empty values will retry indefinitely
+                  "", "", 0, 150,
+                  TimeoutError(
+                      "Unable to determine IP address after 150 retries."
+                      " exit:0 stdout:  stderr: "
+                  )
+             ),
+             (    # only retry until success
+                  ["unparseable", "10.69.10.5 (eth0)\n"], "", 0, 1,
+                  "10.69.10.5"
+             ),
+             (
+                   "10.69.10.5 (eth0)\n", "", 0, 0, "10.69.10.5"
+             ),
+        )
+    )
+    @mock.patch("pycloudlib.lxd.instance.time.sleep")
+    @mock.patch("pycloudlib.lxd.instance.subp")
+    def test_ip_parses_ipv4_output_from_lxc(
+        self, m_subp, m_sleep, stdout, stderr, return_code, sleeps, expected
+    ):
+        """IPv4 output matches specific vm name from `lxc list`.
+
+        Errors are retried and result in TimeoutError on failure.
+        """
+        if isinstance(stdout, list):
+            # Create a list of responses based on stdout being a list
+            m_subp.side_effect = [
+                Result(stdout=out, stderr=stderr, return_code=return_code)
+                for out in stdout
+            ]
+        else:
+            m_subp.return_value = Result(
+                stdout=stdout, stderr=stderr, return_code=return_code
+            )
```

Comment:
If you modified the non-list parameters to be single item lists, could this be simplified to:

```suggestion
        m_subp.side_effect = [
            Result(stdout=out, stderr=stderr, return_code=return_code)
            for out in stdout
        ]
```

?

Reply:
While it felt dirty to stuff different types into the parametrized values, if we try to use side_effect for the mocked subp, we have to bloat the side_effect * sleeps (which is 150) to avoid mock stopiteration issues. That said, I don't like type checking in the unittest. I'll use the one item list for stdouts. and check len to determine if we need to use side_effect vs a single uniform return_value mocked.